### ISSUE TITLE
Is there interest in an alternative FFT implementation supporting Bluestein's algorithm for near-prime sizes?

### ISSUE THREAD

**Issue by mreineck** [numpy/numpy#11885]
If there is interest, I could provide an FFT library implemented in C99 that is based on fftpack, but falls back to Bluestein's algorithm for near-prime lengths. Accuracy should also be better than the original fftpack.

You can have a look at the code at https://gitlab.mpcdf.mpg.de/mtr/pocketfft

I can contribute this under the 3-clause BSD license, but I haven't added the necessary file headers yet.

**Comment by matthew-brett** [comment]
That sounds interesting ...  I guess it would need an expansion of tests to make sure those cases had been properly covered?

**Comment by seberg** [comment]
Don't trust such numpy test parts. This is code that has not recieved significant updates for years and years and the test coverage rules were not always as stringent.
So yes, it should very good test coverage, plus code coverage tools (but numpy CI might help with that).

**Comment by mreineck** [comment]
OK, I just ran a quick coverage test with my own unit tests using "gcov", and the only lines not touched are in the function for reporting a fatal error :)

However I don't have any idea how to integrate this into the numpy CI ... I'm not really a Python person.

**Comment by ncgeib** [comment]
I am strongly in favor of replacing the FFTPACK implementation in numpy. The accuracy for real-valued FFTs and large prime sizes is so bad, in principle, it produces wrong results. At first glance, the pocketfft implementation looks good to me, but I know little of C and the conditions on it required by numpy.

I also note that the same accuracy problem exists in scipy. This issue should be fixed once and for all for both packages. In the long term one FFT implementation should be deprecated and removed.

@mreineck Do you know how the speed compares to the original fftpack and/or numpy's fftpack? I noticed that the C implementation in numpy is slower than the Fortran one in scipy. We should make sure we are at least not slower.

**Comment by matthew-brett** [comment]
If you make a PR with the C code, I can help with the tests.

**Comment by seberg** [comment]
I think we still require C89, but not quite sure. But even if, I guess it should be pretty easy to fix those things.

**Comment by mreineck** [comment]
> If you make a PR with the C code, I can help with the tests.

I'll do that, but I won't be able to provide the Python->C glue.
(Well, I can try, but think this may be a job for experts.)

**Comment by mreineck** [comment]
> I think we still require C89, but not quite sure. But even if, I guess it should be pretty easy to fix those things.

I sure hope not! Changing this code back to C89 would be really painful.

**Comment by matthew-brett** [comment]
Hum - I suppose our problem will likely be Visual C++ 2008.

**Comment by mreineck** [comment]
> @mreineck Do you know how the speed compares to the original fftpack and/or numpy's fftpack? I noticed that the C implementation in numpy is slower than the Fortran one in scipy. We should make sure we are at least not slower.

If there is a difference, pocketfft should be slightly faster than the original fftpack code for highly composite lengths, and very much faster for near-prime lengths.
However, much depends on how the benchmark is actually set up. For example, are the twiddle factor computations included in the FFT run time? How often do we expect an FFT to be executed once its twiddle factors have been computed, etc.

**Comment by mreineck** [comment]
> Hum - I suppose our problem will likely be Visual C++ 2008.

I just read that for this compiler it is possible to just declare the code as being C++, which is fine with the C99 subset I'm using (i.e. I'm not using "complex").

**Comment by matthew-brett** [comment]
I _think_ that VC++ 2008 is our least capable requirement, so I _think_ that will work.  But I'd love to hear from someone who knows this stuff better than me.

**Comment by mreineck** [comment]
I just checked ... code compiles flawlessy as C++98 with g++ (if I #define the `restrict` keyword away).
I don't have access to Microsoft platforms, however.

**Comment by mreineck** [comment]
BTW, an ancestor or this library _was_ actually proposed as a replacement for scipy's FFT in 2013. At the time no conclusion was reached, because scipy uses fftpack not only for FFTs but also for DCTs, DSTs etc., and replacing only half of fftpack didn't seem very attractive. Maybe things have changed by now.

**Comment by rgommers** [comment]
> I think that VC++ 2008 is our least capable requirement, so I think that will work. But I'd love to hear from someone who knows this stuff better than me.

We have one more release (1.16) with Python 2.7 support and then we'll drop it. That's a few months away. From then on the minimum requirements is Python 3.5, where C99 should be fine AFAIK. So I'd recommend to just stay with C99 and we can merge a PR as soon as we've branched 1.16.x

**Comment by rgommers** [comment]
> BTW, an ancestor or this library was actually proposed as a replacement for scipy's FFT in 2013. At the time no conclusion was reached, because scipy uses fftpack not only for FFTs but also for DCTs, DSTs etc., and replacing only half of fftpack didn't seem very attractive. Maybe things have changed by now.

Ah, you found a nice can of worms:) First thoughts:
- For `numpy`, we can replace all of fftpack - we should just do that if it's a major improvement
- For `scipy`, we then probably should replace the functions in `fftpack` that match `numpy` (doesn't have to be done at the same point in time)
- The name `scipy.fftpack` is a bit fishy in itself, and especially when the implementation is not Fftpack anymore - TBD what to do
- We have an issue where we have duplicate functions in numpy and scipy, with sometimes not completely matching signatures. We'd like to resolve that at some point.
- We could address all these issues by introducing `scipy.fft` as a new module, leaving out the functions that are in `numpy.fft`.

**Comment by matthew-brett** [comment]
It sounds like @mreineck found an easy way that might make it possible to compile for VC++ 2008 - in which case - why don't we go for the current release?

Ralf - is there a reason `scipy.fftpack` can't import the numpy versions directly?

I think you are saying we could have `scipy.fft` with matching signatures, and `scipy.fftpack` with the old signatures?

**Comment by rgommers** [comment]
> It sounds like @mreineck found an easy way that might make it possible to compile for VC++ 2008 - in which case - why don't we go for the current release?

If that refers to an untested suggestion to use `g++` for a numpy release, then I'm -10 on that. Way too much work and risk. I'm opposed to changing anything in the compiler setup for Python 2.7 at this point in time. There's no reason not to wait for a couple of months - by the time we have a PR that's reviewed and ready for merge, we may have branched 1.16.x already

> Ralf - is there a reason scipy.fftpack can't import the numpy versions directly?

Signatures don't all match, so can't do that (backwards compatibility)

> I think you are saying we could have scipy.fft with matching signatures, and scipy.fftpack with the old signatures?

yep. But this discussion should move to the scipy tracker/list - either way for numpy it doesn't matter, we can replace all of fftpack.

**Comment by mreineck** [comment]
Pull request created. Have fun!

**Comment by mreineck** [comment]
Just curious: why does the switch to Python 3 imply a change in the acceptable C standard? VC++2008 doesn't care whether it compiles for Python 2 or 3.

**Comment by pv** [comment]
@mreineck: it needs to match what python.org binaries use, https://wiki.python.org/moin/WindowsCompilers

**Comment by njsmith** [comment]
@mreineck until recently, different versions of MSVC used different standard libraries, which means that life is much simpler if you use the same compiler version for all your code. And each python interpreter release uses a specific version of MSVC, so extension modules generally use the same version. For 2.7, that's the 2008 version, since that was current when 2.7 was released; newer pythons fortunately use newer compilers, so we can too.

**Comment by mreineck** [comment]
Ah, I see. Thanks!

**Comment by mreineck** [comment]
I've been running a few performance tests on the C part of numpy's fftpack and pocketfft.
The performance advantage of pocketfft seems pretty significant even for highly composite FFT lengths; I get almost twice the performance in tests with lengths that consist of prime factors 2,3,5,7 and 11.

pocketfft even remains faster than numpy's fftpack if I use the simplest possible pocketfft interface, i.e. if I don't store twiddle factors between individual calls to the same FFT.

So, for the beginning it might be worthwhile just to wrap the minmal interface

```
void cfft_backward_noplan(double c[], size_t length);
void cfft_forward_noplan(double c[], size_t length);
void rfft_backward_noplan(double c[], size_t length);
void rfft_forward_noplan(double c[], size_t length);
```

This could save tons of management and caching code on the Python side.

**Comment by mreineck** [comment]
The PR now has a primitive interface to pocketfft which makes the library importable as "numpy.fft.pocketfft". It presents the same interface as "numpy.fft".

This should be usable for carrying out accuracy tests. Due to the simple interface, performance is still suboptimal, however.

I'm almost 100% certain that the glue code is still broken in several places. Please bear with me, I've never done this before.

**Comment by mreineck** [comment]
Maybe not directly related to this issue, but here is something I've been wondering: would it be possible to ship an interface to FFTW with numpy which is activated whenever numpy detects a suitable libfftw.so on the machine? Since no distribution is involved, I don't see any license conflict.

**Comment by charris** [comment]
> ship an interface to FFTW

Used to be one many years ago, but it was removed. I don't recall if it was in NumPy or SciPy, but the decision was to leave it to third parties.

**Comment by rgommers** [comment]
`pyfftw` monkeypatches scipy (and numpy IIRC), and in scipy we have an explicit regression test to keep that working. don't want a direct interface though, using `pyfftw` should be fine for anyone who wants to use FFTW

**Comment by cycomanic** [comment]
Otherwise does anything speak against using [pretty fast fft](https://bitbucket.org/jpommier/pffft) which is a modernization of fftpack (essentially adding translating fftpack to C and adding SIMD instructions), which has supposedly pretty good performance.

**Comment by grlee77** [comment]
>  pyfftw monkeypatches scipy (and numpy IIRC)

I would say pyFFTW **can** be used to monkeypatch numpy or scipy, but it doesn't do so by default.  I tend to just import the needed functions from `pyfftw.interfaces.numpy_fft`  or `pyfftw.interfaces.scipy_fftpack`.  These two match the API of `numpy.fft` and `scipy.fftpack`, respectively.

**Comment by mreineck** [comment]
> Otherwise does anything speak against using pretty fast fft

Unfortunately it only supports a very limited subset of transform lengths.

**Comment by mreineck** [comment]
Time to close this?

**Comment by rgommers** [comment]
Indeed:)

**Comment by rgommers** [comment]
> Otherwise does anything speak against using pretty fast fft which is a modernization of fftpack (essentially adding translating fftpack to C and adding SIMD instructions), which has supposedly pretty good performance.

I had a quick browse: it's pretty old, not really maintained, and 1-D only so not relevant.

**Comment by mreineck** [comment]
I had some more time to work with the code, and have come up with a (still experimental) Python package based on pocketfft: https://gitlab.mpcdf.mpg.de/mtr/pypocketfft

This package has the following features:
 - same basic features as pocketfft (expecially Bluestein support and accuracy/speed of twiddle factor computation)
 - supports single and double precision transforms
 - full support for multi-D transforms
 - can make use of vector instructions (currently SSE2 and AVX, should be easy to extend) for FFTs with more than one dimension.
 - supports purely real Hartley transforms (https://en.wikipedia.org/wiki/Hartley_transform)

I think this could be a nice candidate for use in numpy and scipy, but there are also a few caveats:
 - The interface is similar, but not identical to numpy.fft; I don't support the `s` argument because it makes the code structure much more complex. If truncation/padding is required, this would have to be done before the FFT call. Normalization is done by passing an explicit normalization factor instead of a string argument; this is nore flexible.
 - The code is heavily templated C++ and uses pybind11 for interfacing; this means that a C++14-capable C++ compiler is required.

I'd be happy to post benchmarks, if anyone is interested. Are there any established multi-D FFT benchmarking tools for Python?

**Comment by rgommers** [comment]
nice @mreineck, looks pretty clean

> The interface is similar, but not identical to numpy.fft; I don't support the s argument because it makes the code structure much more complex. If truncation/padding is required, this would have to be done before the FFT call.

This is a good choice I think, it's kind of weird to have `s` (or `n` in the case of `np.fft.fft`) in the API.

> The code is heavily templated C++ and uses pybind11 for interfacing; this means that a C++14-capable C++ compiler is required.

The Windows situation is a bit unfortunate. C++14 needs MSVC 19.0, which we won't get before Python 3.9 I think. It's possible that compiling extensions with a newer MSVC against a Python compiled with an older MSVC is now feasible, but I haven't seen it tried yet.

> I'd be happy to post benchmarks, if anyone is interested. Are there any established multi-D FFT benchmarking tools for Python?

No benchmarks that I know of. Yes, a benchmark would be nice to see if it's not too much work:)

In SciPy we will very likely get a GSoC project to deliver a backend system this summer. That would allow connecting `pypocketfft` as an option, perhaps on Linux and macOS only.

**Comment by mreineck** [comment]
>This is a good choice I think, it's kind of weird to have `s` (or `n` in the case of `np.fft.fft`) in the API.

Very glad to hear that! The only place where this cannot be entirely avoided is `irfftn`, where the code cannot deduce the size of the last axis unamboguously.

> No benchmarks that I know of. Yes, a benchmark would be nice to see if it's not too much work:)

Here is something very ad hoc. Given a dimensionality and maximum extent of the array, it runs several random-sized FFTs with both `numpy` and `pypocketfft` and plots a histogram of the run-time ratios.

Here are a few results from a laptop with an AVX-capable CPU:
![1d](https://user-images.githubusercontent.com/898507/56499141-611c5780-6505-11e9-90f8-707a9d829aea.png)
![2d](https://user-images.githubusercontent.com/898507/56499142-61b4ee00-6505-11e9-9b92-46fd023e4ebb.png)
![3d](https://user-images.githubusercontent.com/898507/56499143-61b4ee00-6505-11e9-8618-eef0d1b7565a.png)

The speed-ups in the 1D case are not due to vectorization, but due to smaller calling overhead on the Python side. As you can see, some transforms have become slower; I'm not sure yet why that happens (the 1D pocketfft code is more or less unchanged), but I'll investigate.

In higher dimensions the improved performance is explained partially by vectorization and partially by avoiding the explicit array transpositions that numpy currently carries out.

When switching to single precision, the speedup becomes more prominent:

![1d_single](https://user-images.githubusercontent.com/898507/56499346-45658100-6506-11e9-937f-a600d6dcfafa.png)
![2d_single](https://user-images.githubusercontent.com/898507/56499347-45658100-6506-11e9-8fe2-c25239d61a69.png)
![3d_single](https://user-images.githubusercontent.com/898507/56499348-45658100-6506-11e9-9234-0fec4b5ca5b7.png)

Here is the code, so you can test on other architectures:
```
# Don't run this benchmark with numpy<1.17 ... it will probably take ages!

import numpy as np
import pypocketfft
from time import time
import matplotlib.pyplot as plt


def bench_nd_fftn(ndim, nmax, ntry, tp, nrepeat, filename=""):
    res=[]
    for n in range(ntry):
        shp = np.random.randint(1,nmax+1,ndim)
        a=(np.random.rand(*shp) + 1j*np.random.rand(*shp)).astype(tp)
        tmin_np=1e38
        for i in range(nrepeat):
            t0=time()
            b=np.fft.fftn(a)
            t1=time()
            tmin_np = min(tmin_np,t1-t0)
        tmin_pp=1e38
        for i in range(nrepeat):
            t0=time()
            b=pypocketfft.fftn(a)
            t1=time()
            tmin_pp = min(tmin_pp,t1-t0)
        res.append(tmin_pp/tmin_np)
    plt.title("t(pypocketfft / numpy 1.17), {}D, {}, max_extent={}".format(ndim, str(tp), nmax))
    plt.xlabel("time ratio")
    plt.ylabel("counts")
    plt.hist(res,bins="auto")
    if filename != "":
        plt.savefig(filename)
    plt.show()

bench_nd_fftn(1, 8192, 1000, "c16", 10, "1d.png")
bench_nd_fftn(2, 2048, 100, "c16", 2, "2d.png")
bench_nd_fftn(3, 256, 100, "c16", 1, "3d.png")
bench_nd_fftn(1, 8192, 1000, "c8", 10, "1d_single.png")
bench_nd_fftn(2, 2048, 100, "c8", 2, "2d_single.png")
bench_nd_fftn(3, 256, 100, "c8", 1, "3d_single.png")

```

**Comment by insertinterestingnamehere** [comment]
> > The code is heavily templated C++ and uses pybind11 for interfacing; this means that a C++14-capable C++ compiler is required.
> 
> The Windows situation is a bit unfortunate. C++14 needs MSVC 19.0, which we won't get before Python 3.9 I think. It's possible that compiling extensions with a newer MSVC against a Python compiled with an older MSVC is now feasible, but I haven't seen it tried yet.

Minor aside here: I think pybind11 supports MSVC 2015. This may not be a huge issue. From what I've seen in practice, the partial C++14 support there gets you pretty far. Using later compilers is supposed to be possible now too, but I don't know what the details of shipping the runtime libraries, etc. would look like.

**Comment by mreineck** [comment]
That sounds encouraging! I can try to dial back the used C++14 features to a minimum.

One thing I'm using is `aligned_alloc()` (which apparently was only standardized in C++17...), but it should not be a big deal to replace that with `posix_memalign()`, if that is more readily available on Windows.

I definitely will need help with testing on Microsoft platforms, since I'm not using them myself.

**Comment by mreineck** [comment]
I forgot to mention that vectorization is done by means of the gcc vector extension syntax (https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html). I verifed that this works fine with gcc and clang (should also work with icc, but I couldn't test this yet); for MSVC this might have to be disabled, but the scalar part should work fine.

**Comment by insertinterestingnamehere** [comment]
For aligned allocation in MSVC 2015/2017 you need to use `_aligned_malloc` and `_aligned_free`. I'm not sure what to do about the vector extensions though. MSVC has its own vector extensions, but the syntax is different and I haven't used them enough to offer much help there.

**Comment by mreineck** [comment]
Some updates:
- Intel `icpc` apparently doesn't understand all of the gcc vector extensions. I tried with version 19.0, and it was unhappy about mixed simd-scalar operations. So no vector support for Intel at this time.
- I added in-place support to the functions where this makes sense (`fftn`, `ifftn`, `hartley`). If this is selected, the result will be placed into the input array. For 1D transforms this is probably not a big gain, but in the multi-D case it can make a big difference in memory consumption.
